README
An implementation of "Using Uninitialized Memory for Fun and Profit"
http://research.swtch.com/sparse

This is a set implementation for which add(), contains(), remove(), size(),
and clear() operations are all O(1).

Set iteration, intersect, and cloning are O(n).

This performance boost comes by trading memory for speed. Specifically,
this is terribly inefficient in terms of space, because it allocates twice
the amount of memory as required to hold the maxinum number of elements.

Russ Cox, in "Using Uninitialized Memory for Fun and Profit" writes:
  The sparse set is as fast or faster than bit vectors for every operation.
  The only problem is the space cost: two words replace each bit. Still,
  there are times when the speed differences are enough to balance the
  added memory cost. Briggs and Torczon point out that liveness sets used
  during register allocation inside a compiler are usually small and are
  cleared very frequently, making sparse sets the representation of choice.


Dependencies
------------
The tests depend upon libecheck
* https://github.com/ericherman/libecheck


Packaging
---------
autoreconf -iv &&
 ./configure &&
 make &&
 make distcheck &&
 echo "Success."


License
-------
Unless a file specifically says otherwise, all content is Licensed under the
terms of the GNU Lesser General Public License (LGPL), version 3 or later.

See COPYING.GPL3 and COPYING.LGPL3 for details.



NEWS
libfastset NEWS

* Initial Release version 1.0.0 (2016-05-20) [stable]


tests/test-fastset.h
#include <stdio.h>
#include <echeck.h>

#include "../src/fastset.h"

int global_make_valgrind_happy = 0;

struct fastset_s *create_a_fastset(size_t max_value, char *msg)
{
	struct fastset_s *fs;
	int zero_memory_before_use;

	if (max_value == 0) {
		max_value = 1024 * 1024;
	};

	zero_memory_before_use = global_make_valgrind_happy;
	fs = fastset_create(max_value, zero_memory_before_use);
	if (!fs) {
		fprintf(stderr, "%s could not allocate fastset_s (%lu)\n", msg,
			(unsigned long)max_value);
		exit(EXIT_FAILURE);
	}

	return fs;
}

#define TEST_FASTSET_MAIN(func) \
int main(int argc, char **argv) \
{ \
        int failures = 0; \
        if (argc > 1) { \
                global_make_valgrind_happy = atoi(argv[1]); \
        } \
        failures += func; \
        if (failures) { \
                fprintf(stderr, "%d failures in %s \n", failures, __FILE__);\
        } \
        return failures ? 1 : 0; \
}

#define TEST_FASTSET_MAIN2(func1, func2) \
int main(int argc, char **argv) \
{ \
        int failures = 0; \
        if (argc > 1) { \
                global_make_valgrind_happy = atoi(argv[1]); \
        } \
        failures += func1; \
        failures += func2; \
        if (failures) { \
                fprintf(stderr, "%d failures in %s \n", failures, __FILE__);\
        } \
        return failures ? 1 : 0; \
}


tests/test_fastset_clear.c
#include "test-fastset.h"

int test_fastset_clear()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs;

	msg = "test_fastset_clear";
	fs = create_a_fastset(0, msg);

	fastset_add(fs, 5);
	fastset_add(fs, 9);
	fastset_add(fs, 4211);
	fastset_add(fs, 2);

	msg = "test_fastset_clear fastset_size (A)";
	failures += check_size_t_m(fastset_size(fs), 4, msg);

	fastset_clear(fs);

	msg = "test_fastset_clear fastset_size (B)";
	failures += check_size_t_m(fastset_size(fs), 0, msg);

	fastset_free(fs);
	return failures;
}

/* constant time function */
TEST_FASTSET_MAIN(test_fastset_clear())


src/fastset.c
#include "fastset.h"
#include <stdlib.h>
#include <string.h>

struct fastset_s {
	size_t max_value;
	size_t size;
	size_t *sparse;
	size_t *dense;
	char calloc_flag;
};

struct fastset_s *fastset_create(size_t max_value, int zero_mem)
{
	struct fastset_s *fastset;

	fastset = malloc(sizeof(struct fastset_s));
	if (!fastset) {
		return NULL;
	}

	fastset->max_value = max_value;
	fastset->size = 0;
	fastset->calloc_flag = (zero_mem) ? 1 : 0;

	if (fastset->calloc_flag) {
		fastset->sparse = calloc(1 + max_value, (sizeof(size_t)));
	} else {
		fastset->sparse = malloc((1 + max_value) * (sizeof(size_t)));
	}
	if (!fastset->sparse) {
		free(fastset);
		return NULL;
	}

	if (fastset->calloc_flag) {
		fastset->dense = calloc(1 + max_value, (sizeof(size_t)));
	} else {
		fastset->dense = malloc((1 + max_value) * (sizeof(size_t)));
	}
	if (!fastset->dense) {
		free(fastset->sparse);
		free(fastset);
		return NULL;
	}

	return fastset;
}

void fastset_free(struct fastset_s *fastset)
{
	free(fastset->dense);
	free(fastset->sparse);
	free(fastset);
}

int fastset_add(struct fastset_s *fastset, size_t value)
{
	if (value > fastset->max_value) {
		return 0;
	}
	if (fastset_contains(fastset, value)) {
		return 1;
	}
	fastset->dense[fastset->size] = value;
	fastset->sparse[value] = fastset->size;
	fastset->size++;
	return 1;
}

int fastset_contains(struct fastset_s *fastset, size_t value)
{
	if (value > fastset->max_value) {
		return 0;
	}
	if (fastset->sparse[value] >= fastset->size) {
		return 0;
	}
	return fastset->dense[fastset->sparse[value]] == value;
}

void fastset_remove(struct fastset_s *fastset, size_t value)
{
	size_t tmp;

	if (!fastset_contains(fastset, value)) {
		return;
	}

	tmp = fastset->dense[fastset->size - 1];
	fastset->dense[fastset->sparse[value]] = tmp;
	fastset->sparse[tmp] = fastset->sparse[value];
	fastset->size--;
}

size_t fastset_size(struct fastset_s *fastset)
{
	return fastset->size;
}

size_t fastset_max(struct fastset_s *fastset)
{
	return fastset->max_value;
}

void fastset_clear(struct fastset_s *fastset)
{
	fastset->size = 0;
}

void fastset_foreach(struct fastset_s *fastset,
		     void (*func) (size_t each, void *arg), void *arg)
{
	size_t i;

	for (i = 0; i < fastset->size; i++) {
		(*func) (fastset->dense[i], arg);
	}
}

struct fastset_s *fastset_clone(struct fastset_s *fastset)
{
	struct fastset_s *clone;
	size_t i;

	clone = fastset_create(fastset->max_value, fastset->calloc_flag);
	if (!clone) {
		return NULL;
	}

	/* if the set is close to full, just memcpy */
	if ((fastset->max_value / fastset->size) < 4) {
		clone->size = fastset->size;
		memcpy(clone->dense, fastset->dense,
		       sizeof(size_t) * fastset->size);
		memcpy(clone->sparse, fastset->sparse,
		       sizeof(size_t) * (1 + fastset->max_value));
	} else {
		for (i = 0; i < fastset->size; i++) {
			fastset_add(clone, fastset->dense[i]);
		}
	}
	return clone;
}

struct fastset_s *fastset_intersect(struct fastset_s *s1, struct fastset_s *s2)
{
	struct fastset_s *result, *tmp;
	size_t i;

	result = fastset_create(s1->max_value, s1->calloc_flag);
	if (!result) {
		return NULL;
	}

	if (s1->size > s2->size) {
		tmp = s1;
		s1 = s2;
		s2 = tmp;
	}

	for (i = 0; i < s1->size; i++) {
		if (fastset_contains(s2, s1->dense[i])) {
			fastset_add(result, s1->dense[i]);
		}
	}

	return result;
}

struct fastset_s *fastset_union(struct fastset_s *s1, struct fastset_s *s2)
{
	struct fastset_s *result;
	size_t i, max_value;
	int zero_mem;

	max_value =
	    (s1->max_value < s2->max_value) ? s2->max_value : s1->max_value;
	zero_mem = (s1->calloc_flag || s2->calloc_flag);

	result = fastset_create(max_value, zero_mem);
	if (!result) {
		return NULL;
	}

	for (i = 0; i < s1->size; i++) {
		fastset_add(result, s1->dense[i]);
	}
	for (i = 0; i < s2->size; i++) {
		fastset_add(result, s2->dense[i]);
	}

	return result;
}

struct fastset_s *fastset_minus(struct fastset_s *s1, struct fastset_s *s2)
{
	struct fastset_s *result;
	size_t i;

	result = fastset_create(s1->max_value, s1->calloc_flag);
	if (!result) {
		return NULL;
	}

	for (i = 0; i < s1->size; i++) {
		if (!fastset_contains(s2, s1->dense[i])) {
			fastset_add(result, s1->dense[i]);
		}
	}

	return result;
}

struct fastset_s *fastset_unique(struct fastset_s *s1, struct fastset_s *s2)
{
	struct fastset_s *result;
	size_t i, max_value;
	int zero_mem;

	max_value =
	    (s1->max_value < s2->max_value) ? s2->max_value : s1->max_value;
	zero_mem = (s1->calloc_flag || s2->calloc_flag);

	result = fastset_create(max_value, zero_mem);
	if (!result) {
		return NULL;
	}

	for (i = 0; i < s1->size; i++) {
		if (!fastset_contains(s2, s1->dense[i])) {
			fastset_add(result, s1->dense[i]);
		}
	}
	for (i = 0; i < s2->size; i++) {
		if (!fastset_contains(s1, s2->dense[i])) {
			fastset_add(result, s2->dense[i]);
		}
	}

	return result;
}

int fastset_equal(struct fastset_s *s1, struct fastset_s *s2)
{
	size_t i;

	if (s1->size != s2->size) {
		return 0;
	}

	for (i = 0; i < s1->size; i++) {
		if (!fastset_contains(s2, s1->dense[i])) {
			return 0;
		}
	}
	return 1;
}

int fastset_disjoint(struct fastset_s *s1, struct fastset_s *s2)
{
	size_t i;
	struct fastset_s *tmp;

	if (!s1->size && !s2->size) {
		return 1;
	}
	if (s1->size > s2->size) {
		tmp = s2;
		s2 = s1;
		s1 = tmp;
	}

	for (i = 0; i < s1->size; i++) {
		if (fastset_contains(s2, s1->dense[i])) {
			return 0;
		}
	}
	return 1;
}

int fastset_subset(struct fastset_s *s1, struct fastset_s *s2, int strict)
{
	size_t i;

	if (s1->size < s2->size) {
		return 0;
	}

	for (i = 0; i < s2->size; i++) {
		if (!fastset_contains(s1, s2->dense[i])) {
			return 0;
		}
	}
	return (strict && (s1->size == s2->size)) ? 0 : 1;
}

int fastset_superset(struct fastset_s *s1, struct fastset_s *s2, int strict)
{
	return fastset_subset(s2, s1, strict);
}


test-fastset.c
#include <stdio.h>
#include <echeck.h>

#include "fastset.h"

int global_make_valgrind_happy = 0;

struct param_s {
	size_t values[10];
	size_t filled;
};

void fill_array(size_t each, void *arg)
{
	struct param_s *param = (struct param_s *)arg;
	param->values[param->filled++] = each;
}

struct fastset_s *create_a_fastset(size_t max_value, char *msg)
{
	struct fastset_s *fs;

	if (max_value == 0) {
		max_value = 1024 * 1024;
	};

	fs = fastset_create(max_value, global_make_valgrind_happy);
	if (!fs) {
		fprintf(stderr, "%s could not allocate fastset_s (%lu)\n", msg,
			(unsigned long)max_value);
		exit(EXIT_FAILURE);
	}

	return fs;
}

int test_fastset_create()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs;
	size_t max_value = 42;

	msg = "test_fastset_create";
	fs = create_a_fastset(max_value, msg);

	msg = "test_fastset_create fastset_size";
	failures += check_size_t_m(fastset_size(fs), 0, msg);

	msg = "test_fastset_create fastset_max";
	failures += check_size_t_m(fastset_max(fs), max_value, msg);

	fastset_free(fs);
	return failures;
}

int test_fastset_add_size_contains()
{
	int failures = 0;
	char *msg, buf[80];
	struct fastset_s *fs;
	size_t i, max_value = 35;

	msg = "test_fastset_add_size_contains";
	fs = create_a_fastset(max_value, msg);

	msg = "test_fastset_add_size_contains (A)";
	failures += check_size_t_m(fastset_max(fs), max_value, msg);

	/* adding */
	msg = "test_fastset_add_size_contains (B)";
	failures += check_int_m(fastset_add(fs, 3), 1, msg);
	msg = "test_fastset_add_size_contains (C)";
	failures += check_int_m(fastset_add(fs, max_value + 1), 0, msg);

	msg = "test_fastset_add_size_contains (D)";
	failures += check_int_m(fastset_add(fs, 3), 1, msg);
	msg = "test_fastset_add_size_contains (E)";
	failures += check_int_m(fastset_add(fs, 7), 1, msg);
	msg = "test_fastset_add_size_contains (F)";
	failures += check_int_m(fastset_add(fs, 1), 1, msg);

	/* size */
	msg = "test_fastset_add_size_contains (G)";
	failures += check_size_t_m(fastset_size(fs), 3, msg);
	msg = "test_fastset_add_size_contains (H)";
	failures += check_size_t_m(fastset_max(fs), max_value, msg);

	/* contains */
	msg = "test_fastset_add_size_contains (I)";
	failures += check_int_m(fastset_contains(fs, 1), 1, msg);
	msg = "test_fastset_add_size_contains (J)";
	failures += check_int_m(fastset_contains(fs, 2), 0, msg);
	msg = "test_fastset_add_size_contains (K)";
	failures += check_int_m(fastset_contains(fs, 3), 1, msg);
	msg = "test_fastset_add_size_contains (L)";
	failures += check_int_m(fastset_contains(fs, 4), 0, msg);
	msg = "test_fastset_add_size_contains (M)";
	failures += check_int_m(fastset_contains(fs, 7), 1, msg);
	msg = "test_fastset_add_size_contains (N)";
	failures += check_int_m(fastset_contains(fs, 8), 0, msg);

	for (i = max_value + 1; i < max_value + 100000; i++) {
		sprintf(buf, "test_fastset_contains (%lu)", (unsigned long)i);
		failures += check_int_m(fastset_contains(fs, i), 0, msg);
	}

	fastset_free(fs);
	return failures;
}

int test_fastset_remove()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs;

	msg = "test_fastset_remove";
	fs = create_a_fastset(0, msg);

	fastset_add(fs, 1);
	fastset_add(fs, 3);
	fastset_add(fs, 7);

	fastset_remove(fs, 3);

	msg = "test_fastset_remove fastset_size";
	failures += check_size_t_m(fastset_size(fs), 2, msg);

	msg = "test_fastset_remove fastset_contains (A)";
	failures += check_int_m(fastset_contains(fs, 1), 1, msg);
	msg = "test_fastset_remove fastset_contains (B)";
	failures += check_int_m(fastset_contains(fs, 3), 0, msg);
	msg = "test_fastset_remove fastset_contains (C)";
	failures += check_int_m(fastset_contains(fs, 7), 1, msg);

	fastset_free(fs);
	return failures;
}

int test_fastset_clear()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs;

	msg = "test_fastset_clear";
	fs = create_a_fastset(0, msg);

	fastset_add(fs, 5);
	fastset_add(fs, 9);
	fastset_add(fs, 4211);
	fastset_add(fs, 2);

	msg = "test_fastset_clear fastset_size (A)";
	failures += check_size_t_m(fastset_size(fs), 4, msg);

	fastset_clear(fs);

	msg = "test_fastset_clear fastset_size (B)";
	failures += check_size_t_m(fastset_size(fs), 0, msg);

	fastset_free(fs);
	return failures;
}

int test_fastset_foreach()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs;
	size_t i;
	struct param_s *params;
	size_t expected[3] = { 3, 7, 1 };

	params = calloc(1, sizeof(struct param_s));
	if (!params) {
		fprintf(stderr, "woops, could not allocate param_s\n");
		return 1;
	}

	msg = "test_fastset_foreach";
	fs = create_a_fastset(0, msg);

	for (i = 0; i < 3; i++) {
		fastset_add(fs, expected[i]);
	}

	fastset_foreach(fs, fill_array, (void *)params);

	failures +=
	    check_byte_array((unsigned char *)expected, sizeof(size_t) * 3,
			     (unsigned char *)params->values,
			     sizeof(size_t) * params->filled);

	free(params);
	fastset_free(fs);
	return failures;
}

int test_fastset_clone(int full)
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs, *clone;

	msg = "test_fastset_clone";
	fs = create_a_fastset(full ? 3 : 0, msg);

	fastset_add(fs, 0);
	fastset_add(fs, 1);
	fastset_add(fs, 2);
	fastset_add(fs, 3);

	clone = fastset_clone(fs);

	msg = "test_fastset_clone (A)";
	failures += check_size_t_m(fastset_max(fs), fastset_max(clone), msg);

	msg = "test_fastset_clone (B)";
	failures += check_size_t_m(fastset_size(fs), fastset_size(clone), msg);

	msg = "test_fastset_clone (C)";
	failures += check_int_m(fastset_contains(clone, 0), 1, msg);
	msg = "test_fastset_clone (D)";
	failures += check_int_m(fastset_contains(clone, 1), 1, msg);
	msg = "test_fastset_clone (E)";
	failures += check_int_m(fastset_contains(clone, 2), 1, msg);
	msg = "test_fastset_clone (F)";
	failures += check_int_m(fastset_contains(clone, 3), 1, msg);

	fastset_remove(clone, 2);
	msg = "test_fastset_clone (G)";
	failures +=
	    check_size_t_m(fastset_size(fs) - 1, fastset_size(clone), msg);
	msg = "test_fastset_clone (H)";
	failures += check_int_m(fastset_contains(fs, 2), 1, msg);

	fastset_free(clone);
	fastset_free(fs);
	return failures;
}

int test_fastset_intersect()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2, *result;

	msg = "test_fastset_intersect";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	fastset_add(fs1, 0);
	fastset_add(fs1, 1);
	fastset_add(fs1, 2);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	fastset_add(fs1, 5);
	fastset_add(fs1, 6);

	fastset_add(fs2, 2);
	fastset_add(fs2, 3);
	fastset_add(fs2, 5);
	fastset_add(fs2, 7);

	result = fastset_intersect(fs1, fs2);
	if (!result) {
		fprintf(stderr, "whoah, couldn't intersect");
		exit(EXIT_FAILURE);
	}

	msg = "test_fastset_intersect (A)";
	failures += check_int_m(fastset_contains(result, 2), 1, msg);
	msg = "test_fastset_intersect (B)";
	failures += check_int_m(fastset_contains(result, 3), 1, msg);
	msg = "test_fastset_intersect (C)";
	failures += check_int_m(fastset_contains(result, 5), 1, msg);
	msg = "test_fastset_intersect (D)";
	failures += check_int_m(fastset_size(result), 3, msg);

	fastset_free(result);
	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_union()
{
	int failures = 0;
	char *msg;
	char buf[80];
	struct fastset_s *fs1, *fs2, *result;
	size_t i;

	msg = "test_fastset_intersect";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	fastset_add(fs1, 0);
	fastset_add(fs1, 1);
	fastset_add(fs1, 2);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	fastset_add(fs1, 5);
	fastset_add(fs1, 6);

	fastset_add(fs2, 2);
	fastset_add(fs2, 3);
	fastset_add(fs2, 5);
	fastset_add(fs2, 7);

	result = fastset_union(fs1, fs2);
	if (!result) {
		fprintf(stderr, "whoah, couldn't union");
		exit(EXIT_FAILURE);
	}
	msg = "test_fastset_union size";
	failures += check_size_t_m(fastset_size(result), 8, msg);

	for (i = 0; i < 8; i++) {
		sprintf(buf, "test_fastset_union (%lu)", (unsigned long)i);
		failures += check_int_m(fastset_contains(result, i), 1, msg);
	}

	fastset_free(result);
	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_minus()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2, *result;
	size_t i;
	char buf[80];
	size_t expected[4] = { 0, 1, 4, 6 };

	msg = "test_fastset_minus";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	fastset_add(fs1, 0);
	fastset_add(fs1, 1);
	fastset_add(fs1, 2);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	fastset_add(fs1, 5);
	fastset_add(fs1, 6);

	fastset_add(fs2, 2);
	fastset_add(fs2, 3);
	fastset_add(fs2, 5);
	fastset_add(fs2, 7);

	result = fastset_minus(fs1, fs2);
	if (!result) {
		fprintf(stderr, "whoah, couldn't minus");
		exit(EXIT_FAILURE);
	}

	msg = "test_fastset_minus size";
	failures += check_size_t_m(fastset_size(result), 4, msg);

	for (i = 0; i < 4; i++) {
		sprintf(buf, "test_fastset_minus (%lu)", (unsigned long)i);
		failures +=
		    check_int_m(fastset_contains(result, expected[i]), 1, msg);
	}

	fastset_free(result);
	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_unique()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2, *result;
	size_t i;
	char buf[80];
	size_t expected[5] = { 0, 1, 4, 6, 7 };

	msg = "test_fastset_minus";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	fastset_add(fs1, 0);
	fastset_add(fs1, 1);
	fastset_add(fs1, 2);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	fastset_add(fs1, 5);
	fastset_add(fs1, 6);

	fastset_add(fs2, 2);
	fastset_add(fs2, 3);
	fastset_add(fs2, 5);
	fastset_add(fs2, 7);

	result = fastset_unique(fs1, fs2);
	if (!result) {
		fprintf(stderr, "whoah, couldn't unique");
		exit(EXIT_FAILURE);
	}
	msg = "test_fastset_unique size";
	failures += check_size_t_m(fastset_size(result), 5, msg);

	for (i = 0; i < 5; i++) {
		sprintf(buf, "test_fastset_unique (%lu)", (unsigned long)i);
		failures +=
		    check_int_m(fastset_contains(result, expected[i]), 1, msg);
	}

	fastset_free(result);
	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_equal()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2;

	msg = "test_fastset_equal";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	failures += check_int_m(fastset_equal(fs1, fs2), 1, msg);

	fastset_add(fs1, 1);
	msg = "test_fastset_equal [1],[]";
	failures += check_int_m(fastset_equal(fs1, fs2), 0, msg);

	fastset_add(fs2, 1);
	msg = "test_fastset_equal [1],[1]";
	failures += check_int_m(fastset_equal(fs1, fs2), 1, msg);

	fastset_add(fs2, 2);
	msg = "test_fastset_equal [1],[1,2]";
	failures += check_int_m(fastset_equal(fs1, fs2), 0, msg);

	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_disjoint()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2;

	msg = "test_fastset_disjoint";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	failures += check_int_m(fastset_disjoint(fs1, fs2), 1, msg);

	fastset_add(fs1, 1);
	msg = "test_fastset_disjoint [1],[]";
	failures += check_int_m(fastset_disjoint(fs1, fs2), 1, msg);

	fastset_add(fs2, 1);
	msg = "test_fastset_disjoint [1],[1]";
	failures += check_int_m(fastset_disjoint(fs1, fs2), 0, msg);

	fastset_add(fs2, 2);
	msg = "test_fastset_disjoint [1],[1,2]";
	failures += check_int_m(fastset_disjoint(fs1, fs2), 0, msg);

	fastset_clear(fs1);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	msg = "test_fastset_disjoint [3,4],[1,2]";
	failures += check_int_m(fastset_disjoint(fs1, fs2), 1, msg);

	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int test_fastset_subset_superset()
{
	int failures = 0;
	char *msg;
	struct fastset_s *fs1, *fs2;

	msg = "test_fastset_subset";
	fs1 = create_a_fastset(10, msg);
	fs2 = create_a_fastset(10, msg);

	msg = "test_fastset_subset [],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_superset [],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_subset strict [],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 0, msg);
	msg = "test_fastset_superset strict [],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 0, msg);

	fastset_add(fs1, 1);
	msg = "test_fastset_subset [1],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_superset [1],[]";
	failures += check_int_m(fastset_superset(fs1, fs2, 0), 0, msg);
	msg = "test_fastset_subset strict [1],[]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 1, msg);
	msg = "test_fastset_superset strict [1],[]";
	failures += check_int_m(fastset_superset(fs1, fs2, 1), 0, msg);
	msg = "test_fastset_subset [],[1]";
	failures += check_int_m(fastset_subset(fs2, fs1, 0), 0, msg);
	msg = "test_fastset_superset [],[1]";
	failures += check_int_m(fastset_superset(fs2, fs1, 0), 1, msg);
	msg = "test_fastset_subset strict [],[1]";
	failures += check_int_m(fastset_subset(fs2, fs1, 1), 0, msg);
	msg = "test_fastset_superset strict [],[1]";
	failures += check_int_m(fastset_superset(fs2, fs1, 1), 1, msg);

	fastset_add(fs2, 1);
	msg = "test_fastset_subset [1],[1]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_superset [1],[1]";
	failures += check_int_m(fastset_superset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_subset strict [1],[1]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 0, msg);
	msg = "test_fastset_superset strict [1],[1]";
	failures += check_int_m(fastset_superset(fs1, fs2, 1), 0, msg);

	fastset_add(fs2, 2);
	msg = "test_fastset_subset [1],[1,2]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 0, msg);
	msg = "test_fastset_superset [1],[1,2]";
	failures += check_int_m(fastset_superset(fs1, fs2, 0), 1, msg);
	msg = "test_fastset_subset strict [1],[1,2]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 0, msg);
	msg = "test_fastset_superset strict [1],[1,2]";
	failures += check_int_m(fastset_superset(fs1, fs2, 1), 1, msg);
	msg = "test_fastset_subset [1,2], [1]";
	failures += check_int_m(fastset_subset(fs2, fs1, 0), 1, msg);
	msg = "test_fastset_superset [1,2], [1]";
	failures += check_int_m(fastset_superset(fs2, fs1, 0), 0, msg);
	msg = "test_fastset_subset strict [1,2], [1]";
	failures += check_int_m(fastset_subset(fs2, fs1, 1), 1, msg);
	msg = "test_fastset_superset strict [1,2], [1]";
	failures += check_int_m(fastset_superset(fs2, fs1, 1), 0, msg);

	fastset_clear(fs1);
	fastset_add(fs1, 3);
	fastset_add(fs1, 4);
	msg = "test_fastset_subset [3,4],[1,2]";
	failures += check_int_m(fastset_subset(fs1, fs2, 0), 0, msg);
	msg = "test_fastset_superset [3,4],[1,2]";
	failures += check_int_m(fastset_superset(fs1, fs2, 0), 0, msg);
	msg = "test_fastset_subset strict [3,4],[1,2]";
	failures += check_int_m(fastset_subset(fs1, fs2, 1), 0, msg);
	msg = "test_fastset_superset strict [3,4],[1,2]";
	failures += check_int_m(fastset_superset(fs1, fs2, 1), 0, msg);

	fastset_free(fs2);
	fastset_free(fs1);
	return failures;
}

int main(int argc, char *argv[])
{
	int failures = 0;

	if (argc > 1) {
		global_make_valgrind_happy = atoi(argv[1]);
	}

	/* constant time functions */
	failures += test_fastset_create();
	failures += test_fastset_add_size_contains();
	failures += test_fastset_remove();
	failures += test_fastset_clear();

	/* O(n) functions */
	failures += test_fastset_foreach();
	failures += test_fastset_clone(0);
	failures += test_fastset_clone(1);
	failures += test_fastset_intersect();
	failures += test_fastset_union();
	failures += test_fastset_minus();
	failures += test_fastset_unique();
	failures += test_fastset_equal();
	failures += test_fastset_disjoint();
	failures += test_fastset_subset_superset();

	if (failures) {
		fprintf(stderr, "%d failures in total\n", failures);
	}
	return failures;
}
